### DISCUSSION TITLE
how to derive Write[Either[T,U]]

### DISCUSSION THREAD

**Discussion by berserk-fan** [tpolecat/doobie#1838] [Q&A] [answered]
Is it possible to derive Write[Either[T,U]] from Write[T] and Write[U]?

**Comment by jatcwang** [comment, marked as answer]
Not generically no. How do we know wheter a result column contains T or U?

What might work for you use to have a sealed trait / enum, and deriving JSON encoder/decoders for it. Then use the doobie's circe integration to read the column. For an example see https://stackoverflow.com/questions/67789229/how-i-can-insert-case-object-as-jsonb-format-via-doobie

**Reply by berserk-fan** [reply]
> How do we know wheter a result column contains T or U?
I guess we may know it when we have the actual value.

for example toFragment(t: Either[T,U], sql: String) may pattern match on t and use appropriate writer.

also I need it because for my specific use case I prefer T Either U Either ... instead of sealed trait

thanks for the answer, I can handle it without generic derivation of Writer[Either[T,U]]

**Reply by jatcwang** [reply]
I see what you mean - you only want a Write[Either[T,U]]. I think you'll need to put in some safe guards too because T and U can have different number of fields which means you can't use the same SQL right? (unless your SQL can work with different number of parameters).

So yeah not sure there's enough general use case to warrant an addition to Doobie itself, but you should be able to write a Write[Either[T, U]] in your own project. Let me know if you need help :)
